src/librustc_driver/driver.rs

```diff
@@ -1335,9 +1335,10 @@ pub fn build_output_filenames(input: &Input,
                                            .values()
                                            .filter(|a| a.is_none())
                                            .count();
-            let ofile = if unnamed_output_types > 1 {
-                sess.warn("ignoring specified output filename because multiple outputs were \
-                           requested");
+            let ofile = if unnamed_output_types > 1 &&
+                            sess.opts.output_types.contains_key(&OutputType::Exe) {
+                sess.warn("ignoring output name requested with -o for \"link\" output because \
```

Comment:
I think you could change the error message less - in particular, keeping `filename` rather than `name` seems more informative, and I don't think mentioning `-o` is necessary. Nit: I would prefer using single quotes or backticks rather than double quotes around `link`.